youtube_dl/extractor/viu.py

```diff
@@ -223,20 +269,64 @@ def _real_extract(self, url):
         if not video_data:
             raise ExtractorError('This video is not available in your region.', expected=True)
 
-        stream_data = self._download_json(
-            'https://d1k2us671qcoau.cloudfront.net/distribute_web_%s.php' % country_code,
-            video_id, 'Downloading stream info', query={
-                'ccs_product_id': video_data['ccs_product_id'],
-            }, headers={
-                'Referer': url,
-                'Origin': re.search(r'https?://[^/]+', url).group(0),
-            })['data']['stream']
+        duration_limit = False
+        query = {
+            'ccs_product_id': video_data['ccs_product_id'],
+            'language_flag_id': self._LANGUAGE_FLAG.get(lang_code.lower()) or '3',
+        }
+        headers = {
+            'Referer': re.search(r'https?://[^/]+', url).group(0),
+            'Origin': re.search(r'https?://[^/]+', url).group(0),
+        }
+        try:
+            stream_data = self._download_json(
+                'https://d1k2us671qcoau.cloudfront.net/distribute_web_%s.php' % country_code,
+                video_id, 'Downloading stream info', query=query, headers=headers)
+            stream_data = self._detect_error(stream_data)['stream']
+        except (ExtractorError, KeyError):
+            stream_data = None
+            if video_data.get('user_level', 0) > 0:
+                user = self._login(country_code, video_id)
```

Comment:
You could use
```
def _real_initialize(self):
        self._login()
```
above line 230 so before defining `_login()`

Have a look here how it's done
https://github.com/blackjack4494/youtube-dlc/blob/7eff09d332745e48acf657fbb65be40680d764f2/youtube_dlc/extractor/viki.py#L88-L107

You need some `_USER = None` or so at the top of the extractor then

That way login will automatically executed whenever you supply credentials.

Reply:
i wrote like this since I don't want to login unless really necessary. @zackmark29 has once mentioned the server counts the login sessions such that one user cannot login into more a limited devices at the same time. is it not the common way to do or should I make it always login if credentials is provided?